### PR TITLE
Replace PGI by NVHPC in CMake setup

### PR DESCRIPTION
**Pull request by masterleinad**
`CMake` identifies `nvc++` as `NVHPC` for `CMake` >=3.20 and as PGI before that. Since nvc++ is supposed to be backward-compatible with pgc++ just setting `NVHPC` as internal compiler identification for `PGI` should be fine and avoids distinguishing between the two.

### CODE REVIEW

File: core/src/Cuda/Kokkos_Cuda_Half.hpp [hunk 1]
```diff
@@ -315,8 +315,8 @@ class half_t {
     // Use non-volatile val_ref to suppress:
     // "warning: implicit dereference will not access object of type ‘volatile
     // __half’ in statement"
-    auto val_ref = const_cast<impl_type&>(val);
-    val_ref      = __float2half(__half2float(const_cast<impl_type&>(val)) +
+    auto& val_ref = const_cast<impl_type&>(val);
+    val_ref       = __float2half(__half2float(const_cast<impl_type&>(val)) +
                            __half2float(rhs.val));
```
Comment (crtrott) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:320]: omg this operator was simply wrong before. val_ref was a value type wasn't it? Not a reference type? Can we get a test for this?

**Review by crtrott — APPROVED** [review]
I want a test written for the volatile operators of half. @e10harvey Note that the volatile operators where this PR changes to auto& didn't actually update val I believe.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp (hunk 1 above)
Comment (dalg24) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:320]: Does not belong to this PR IMO

File: cmake/kokkos_compiler_id.cmake [hunk 2]
```diff
@@ -165,6 +165,7 @@ ELSEIF(KOKKOS_CXX_COMPILER_ID STREQUAL PGI)
   IF(KOKKOS_CXX_COMPILER_VERSION VERSION_LESS 17.4)
     MESSAGE(FATAL_ERROR "${KOKKOS_MESSAGE_TEXT}")
   ENDIF()
+  SET(KOKKOS_CXX_COMPILER_ID NVHPC CACHE STRING INTERNAL FORCE)
```
Comment (dalg24) [inline cmake/kokkos_compiler_id.cmake:168]: Add a comment here.

**Review by dalg24 — CHANGES_REQUESTED** [review]

File: core/src/Cuda/Kokkos_Cuda_Half.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:320]: Moved to #4198 and asked @e10harvey for a test.

File: cmake/kokkos_compiler_id.cmake (hunk 2 above)
Comment (masterleinad) [reply cmake/kokkos_compiler_id.cmake:171]: Done.

**Comment by masterleinad** [comment]
Retest this please.

**Review by dalg24 — APPROVED** [review]

File: core/src/Cuda/Kokkos_Cuda_Half.hpp (hunk 1 above)
Comment (e10harvey) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp]: These tests are part of #4101 which is ready for another round of reviews. See https://github.com/kokkos/kokkos/pull/4101/files#diff-6980b46765a569d9ae64dffdbe6a0db53b9ed49c8b9157236055a90add5b47afR277-R352.

**Comment by e10harvey** [comment]
> I want a test written for the volatile operators of half. @e10harvey Note that the volatile operators where this PR changes to auto& didn't actually update val I believe.

These tests are part of #4101.

**Pull request merged**